### PR TITLE
Initial optimistic locking implmentation.

### PR DESCRIPTION
**Pull request by davidtsai**
_Thanks for wanting to fix something on Sequelize - we already love you long time! Please delete this text and fill in the template below. If unsure about something, just do as best as you're able._

_If your PR only contains changes to documentation, you may skip the template below._
### Pull Request check-list

_Please make sure to review and check all of these items:_
- [ ] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Have you added an entry under `Future` in the changelog?

_NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._
### Description of change

_Please provide a description of the change here._

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -200,6 +200,14 @@ var addDefaultAttributes = function() {
       _autoGenerated: true
     };
   }
+  if (this._versionAttribute) {
+    tail[this._versionAttribute] = {
+      type: DataTypes.INTEGER,
+      allowNull: false,
+      defaultValue: 0,
```
Comment (igrayson) [inline lib/model.js]: imo require that we provide `0` during insert

**Comment by alexturek** [comment]
I have a question - why a `version` integer field instead of using `updatedAt` timestamp if that's part of the model?

**Comment by davidtsai** [comment]
Because server clocks aren't that accurate and you could accidentally
clobber an object when updatedAt is equal but one of the writes is actually
stale. In practice this is unlikely to happen at our scale but the ORM
should have a theoretically sound implementation.

**Comment by alexturek** [comment]
I don't understand the clock comment.

1. I get `updatedAt: "2016-09-21T18:11:44.203Z"` on reading the instance.
2. I update the field `foo = 5` on the instance.
3. I optimistically update via `UPDATE things SET foo = 5, updated_at = now() where updated_at = '2016-09-21T18:11:44.203Z'`

**Comment by davidtsai** [comment]
Because optimistic locking works across application hosts eg:
1. Host 1 reads object
2. Host 2 reads object
3. Host 1 changes object, sets updated at to X
4. Host 1 saves
5. Host 2 changes object, sets updated at to same X
6. Host 2 saves and clobbers host 1's changes.

File: lib/model.js (hunk 1 above)
Comment (davidtsai) [inline lib/model.js]: Are you suggesting the model user has to specify the version number on instance creation?  Eg. `Shipment.create({version:0})`?

That seems unnecessary to me.  The user is only going to pass zero everytime.

**Comment by alexturek** [comment]
> Also sequelize doesn't use the DB's clock.

Sequelize you're killing me

**Pull request closed without merging**